src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5042,6 +5042,279 @@ class StubGenerator: public StubCodeGenerator {
+      __ vid_v(vtemp1);
+      __ vrsub_vx(vtable_32, vtemp1, temp1);
+      // vtable_32 group now contains { 0x20, 0x1f, 0x1e, ..., 0x3, 0x2, 0x1 }
+    }
+
+    __ vsetivli(temp0, 16, Assembler::e8, Assembler::m1);
+    if (MaxVectorSize > 16) {
+      // Need to generate vtable_16 explicitly
+      __ mv(temp1, 16);
+      __ vid_v(vtemp1);
+      __ vrsub_vx(vtable_16, vtemp1, temp1);
+      // vtable_16 group now contains { 0x10, 0xf, 0xe, ..., 0x3, 0x2, 0x1 }
+    }
+
+    __ vmv_v_i(vzero, 0);
+
+    __ mv(base, BASE);
+    __ mv(nmax, NMAX);
+
+    // s1 is initialized to the lower 16 bits of adler
+    // s2 is initialized to the upper 16 bits of adler
+    __ srliw(s2, adler, 16); // s2 = ((adler >> 16) & 0xffff)
+    __ zero_extend(s1, adler, 16); // s1 = (adler & 0xffff)
+
+    // The pipelined loop needs at least 16 elements for 1 iteration
+    // It does check this, but it is more effective to skip to the cleanup loop
+    __ mv(temp0, step_16);
+    __ bgeu(len, temp0, L_nmax);
+    __ beqz(len, L_combine);
+
+  __ bind(L_simple_by1_loop);
+    __ lbu(temp0, Address(buff, 0));
+    __ addi(buff, buff, step_1);
+    __ add(s1, s1, temp0);
+    __ add(s2, s2, s1);
+    __ sub(len, len, step_1);
+    __ bgtz(len, L_simple_by1_loop);
+
+    // s1 = s1 % BASE
+    __ remuw(s1, s1, base);
+    // s2 = s2 % BASE
+    __ remuw(s2, s2, base);
+
+    __ j(L_combine);
+
+  __ bind(L_nmax);
+    __ sub(len, len, nmax);
+    __ sub(count, nmax, 16);
+    __ bltz(len, L_by16);
+
+  // Align L_nmax loop by 64
+  __ bind(L_nmax_loop_entry);
+    __ sub(count, count, 32);
+
+  __ bind(L_nmax_loop);
+    adler32_process_bytes(buff, s1, s2, vtable_64, vzero,
+      vbytes, vs1acc, vs2acc, temp0, temp1, temp2, temp3,
+      vtemp1, vtemp2, step_64, Assembler::m4);
+    __ sub(count, count, step_64);
+    __ bgtz(count, L_nmax_loop);
+
+    // There are three iterations left to do
+    adler32_process_bytes(buff, s1, s2, vtable_32, vzero,
+      vbytes, vs1acc, vs2acc, temp0, temp1, temp2, temp3,
+      vtemp1, vtemp2, step_32, Assembler::m2);
+    adler32_process_bytes(buff, s1, s2, vtable_16, vzero,
+      vbytes, vs1acc, vs2acc, temp0, temp1, temp2, temp3, 
+      vtemp1, vtemp2, step_16, Assembler::m1);
+
+    // s1 = s1 % BASE
+    __ remuw(s1, s1, base);
+    // s2 = s2 % BASE
+    __ remuw(s2, s2, base);
+
+    __ sub(len, len, nmax);
+    __ sub(count, nmax, 16);
+    __ bgez(len, L_nmax_loop_entry);
+
+  __ bind(L_by16);
+    __ add(len, len, count);
+    __ bltz(len, L_by1);
+    // Trying to unroll
+    __ mv(count, step_64);
+    __ blt(len, count, L_by16_loop);
+
+  __ bind(L_by16_loop_unroll);
+    adler32_process_bytes(buff, s1, s2, vtable_64, vzero,
+      vbytes, vs1acc, vs2acc, temp0, temp1, temp2, count,
+      vtemp1, vtemp2, step_64, Assembler::m4);
+    __ sub(len, len, step_64);
+    // By now the count should still be 64
+    __ bge(len, count, L_by16_loop_unroll);
+    __ mv(count, step_16);
+    __ blt(len, count, L_by1);
```

Comment:
Question: Why do we need this 'blt' branch after the loop unroll here? The `len` has been subtracted by 16 at `L_by16` by `__ add(len, len, count)` where the input `len == len - nmax` and `count == nmax - 16`.

Reply:
`L_by16` loop just means that the length is less than `NMAX` (which is 5552). So even after subtracting It is possible for `len` to be less than 16 but greater than 0 on this `blt`.

For example, take 90 as an initial `len`. We go to `L_nmax` firstly, bypassing the `L_simple_by1_loop`. Then we branch onto `L_by16` since we're smaller than `nmax`. `add(len, len, count)` instruction means that we have 76 at `len` at this point. After executing `adler32_process_bytes` and `sub((len, len, step_64)` we got 12 in the `len` (smaller than 64), meaning that we have to choose whether we need to process 16 bytes at a step in `L_by16_loop` or go to `L_by1`. And the execution goes to `L_by1` instead of falling through.

Please correct me If you see some cases that doesn't fit into this model

Reply:
So the control flow finally goes to `L_by1` in your case. As you see, there is a `__ add(len, len, 15)` at `L_by1`. Adding 15 and 12, we have 27 in len which is bigger than 16. So why not fall through to `L_by16_loop` in this case? Seems better in performance as we do `adler32_process_bytes` for each 16-bytes at a time.

Reply:
Sorry for such a late reply. It is a good catch, I see it, thanks! However, it is not correct for all cases: consider having -63 after `L_by16_loop_unroll`. In current case (the correct one) we would just jump to `L_by1` and to `L_do_mod`, ending the execution. With unconditional fallthrough the excessive job would have been performed, which is not correct. So as far as I can see the best option would be just to add another condition for `L_by1`, checking for proper `len` sizes and performing `adler32_process_bytes` when needed. Please correct me if I'm wrong

Reply:
> Sorry for such a late reply. It is a good catch, I see it, thanks! However, it is not correct for all cases: consider having -63 after `L_by16_loop_unroll`

Hmm ... I don't understand how this could happen. I see `len >= 64` is ensured when we first enter `L_by16_loop_unroll`. And you do `len -= 64` and compare the result with 64 at the end of each iteration. So I think the final `len` when we exit this loop will be >= 0, right?

BTW: Could you please avoid use `count` as scratch register when say compare `len` with sizes like step_16/64. It's better to uses others like temp0-3.

Reply:
Ohh, I'm sorry, that's my mistake. The `len` is >= 0 after `L_by16_loop_unroll`, you're right. So I removed the `blt` from `L_by16_loop_unroll` end and excessive `adler32_process_bytes` (introduced in [this](https://github.com/openjdk/jdk/pull/18382/commits/0ce230a4ecf1ca81d5f615e42e9758b6be36f498) commit) from `L_by1`, and ran a `test/hotspot/jtreg/compiler/intrinsics/zip/TestAdler32.java` test, which is OK. Thanks for your patience!

Also I used `t2` instead of a `count` for sizes comparisons